include/boost/algorithm/searching/aho_corasick.hpp

```diff
@@ -0,0 +1,283 @@
+/*
+  Copyright (c) Alexander Zaitsev <zamazan4ik@gmail.com>, 2016
+  Distributed under the Boost Software License, Version 1.0. (See
+  accompanying file LICENSE_1_0.txt or copy at
+  http://www.boost.org/LICENSE_1_0.txt)
+  See http://www.boost.org/ for latest version.
+*/
+
+#ifndef BOOST_ALGORITHM_AHO_CORASICK_HPP
+#define BOOST_ALGORITHM_AHO_CORASICK_HPP
+
+#include <vector>
+#include <queue>
+#include <functional>
+
+#include <boost/range/begin.hpp>
+#include <boost/range/end.hpp>
+#include <boost/container/map.hpp>
+#include <boost/unordered_map.hpp>
+
+namespace boost { namespace algorithm {
+
+template <typename T, template<typename ...> class Container, typename ...Args>
+class aho_corasick
+{
+private:
+    class node
+    {
+    public:
+        Container<T, node, Args...> links;
+        node *fail, *term;
+        std::vector<size_t> pat;
+
+        node(node* fail_node = nullptr)
+                : fail(fail_node), term(nullptr)
+        { }
+
+        node* getLink(const T& c)
+        {
+            auto iter = links.find(c);
+            return iter != links.end() ? &iter->second : nullptr;
+        }
+
+        bool isTerminal() const
+        {
+            return !pat.empty();
+        }
+    };
+public:
+    using value_type = T;
+    using node_type = node;
+private:
+    node_type root;
+    node_type* current_state;
+    size_t countStrings = 0;
+    bool isInited = false;
+public:
+    aho_corasick(){}
+
+    template<typename ForwardIterator>
+    explicit aho_corasick(ForwardIterator patBegin, ForwardIterator patEnd)
+    {
+        while(patBegin != patEnd)
+        {
+            insert(*patBegin);
+            ++patBegin;
+        }
+    }
+
+    template<typename Range>
+    explicit aho_corasick(const Range& range) : aho_corasick(boost::begin(range), boost::end(range)) {}
+
+    /// \fn insert(const Range& range)
+    /// \brief Insert pattern in trie
+    ///
+    /// \param range The pattern range
+    ///
+    template<typename Range>
+    void insert(const Range& range)
+    {
+        insert(boost::begin(range), boost::end(range));
+    }
+
+    /// \fn insert(ForwardIterator begin, ForwardIterator end)
+    /// \brief Insert pattern in trie
+    ///
+    /// \param begin The start of the pattern
+    /// \param end   One past the end of the pattern
+    ///
+    template<typename ForwardIterator>
+    void insert(ForwardIterator begin, ForwardIterator end)
+    {
+        isInited = false;
+        size_t patLen = 0;
+        node_type* current_node = &root;
+        for(auto it = begin; it != end; ++it)
+        {
+            ++patLen;
+            node_type* child_node = current_node->getLink(*it);
+            if (!child_node)
+            {
+                node new_node;
+                current_node->links[*it] = std::move(new_node);
+                child_node = &current_node->links[*it];
```

Comment:
I believe that lines 102-103 can be simplified to one line, `current_node->links[*it] = node();`, and that move semantics will still take place in C++11.